File: JavaScript/Stack/20_Valid_Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    if (s.length % 2 !== 0) return false;
    let stack = [];
    for (let i = 0; i < s.length; i++) {
        if (['(', '{', '['].includes(s.charAt(i))) {
            stack.push(s.charAt(i));
        } else {
            if (stack.length === 0) return false;
            if (!validType(stack.pop(), s.charAt(i))) return false;
        }
    }
    return stack.length === 0 ? true : false;
};

const validType = (a, b) => {
    if (a === '(' && b === ')') return true;
    if (a === '{' && b === '}') return true;
    if (a === '[' && b === ']') return true;
    return false;
}

File: Java/Stack/20_Valid_Parentheses.java
class Solution {
    public boolean isPair(char c1, char c2) {
        return (c1 == '(' && c2 == ')') || (c1 == '{' && c2 == '}') || (c1 == '[' && c2 == ']');
    }
    public boolean isValid(String s) {
        Stack < Character > st = new Stack < Character > ();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[') {
                st.push(s.charAt(i));
            } else {
                if (!st.isEmpty() && isPair(st.peek(), s.charAt(i))) {
                    st.pop();
                } else {
                    return false;
                }
            }
        }
        return st.isEmpty();
    }
}